### ISSUE TITLE
Terraform looping infinitely while creating the pool in the load balancer

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-openstack#56]
_This issue was originally opened by @v-2vikc as hashicorp/terraform#15619. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

# PROVIDER ISSUES

**PLEASE NOTE:** Terraform has split out the builtin Providers into their own repositories. For any Provider issues, please open all issues and pull requests in the corresponding repository. An index of supported Providers can be found here:

- https://github.com/terraform-providers

All other issues (that appear to affect multiple or all providers) may be an issue with Terraform's core, and should be opened here. 

------

Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

### Terraform Version
Terraform v0.9.11

### Terraform Configuration Files

provider "openstack" {
  user_name   = "******"
  tenant_name = "********"
  password    = "*******"
  auth_url    = "*********"
 domain_name = "***"
 }

resource "openstack_lb_pool_v2" "test" {
  region = "eu-de-1"
  name = "fix"
  description = "fix-pool"
  protocol    = "HTTP"
  lb_method   = "ROUND_ROBIN"
  listener_id = "4bf97f1a-eb43-46bd-b08d-7582a765324d"

provisioner "local-exec" {
command = "sleep 5"
        }

}

### Debug Output

test> sudo terraform plan

Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed. Cyan entries are data sources to be read.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

+ openstack_lb_pool_v2.test
    admin_state_up: "true"
    description:    "new-pool"
    lb_method:      "ROUND_ROBIN"
    listener_id:    "4bf97f1a-eb43-46bd-b08d-7582a765324d"
    name:           "new"
    protocol:       "HTTP"
    region:         "eu-de-1"
    tenant_id:      "<computed>"

Plan: 1 to add, 0 to change, 0 to destroy.
deployer@pc12rundeck01cc:/tmp/vikram/test> sudo vi test.tf
deployer@pc12rundeck01cc:/tmp/vikram/test> sudo terraform apply
openstack_lb_pool_v2.test: Creating...
  admin_state_up: "" => "true"
  description:    "" => "fix-pool"
  lb_method:      "" => "ROUND_ROBIN"
  listener_id:    "" => "4bf97f1a-eb43-46bd-b08d-7582a765324d"
  name:           "" => "fix"
  protocol:       "" => "HTTP"
  region:         "" => "eu-de-1"
  tenant_id:      "" => "<computed>"
openstack_lb_pool_v2.test: Still creating... (10s elapsed)
openstack_lb_pool_v2.test: Still creating... (20s elapsed)

### Panic Output
NA

### Expected Behavior

As per plan it should only create on pool. 

### Actual Behavior

It is looping infinitely and creating multiple pools.. untill the quota is exceeded in open stack.  

### Steps to Reproduce

1. `terraform apply`

### Important Factoids
Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? Custom version of OpenStack? Tight ACLs?

NA

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- GH-1234

**Comment by jtopjian** [comment]
@v-2vikc This sounds like it's an issue specific to the OpenStack cloud you are using. You might be able to find more information by running the following:

```shell
$ TF_LOG=DEBUG OS_DEBUG=1 terraform apply
```

**Comment by sirajudheenam** [comment]
I have similar issue that keep creating `load balancer pool` until the quota exceeds every 10 seconds and this happens when I have protocol mismatch and neutron kept complaining about this mismatch. 
eg., listener protocol is `HTTP` and pool protocol is `TCP`. 

```hcl
#LISTENER 2
resource "openstack_lb_listener_v2" "listener_2" {
  name = "listener02"
  description = "This is the listener 2 description"
  protocol        = "HTTP"
  protocol_port   = 8081
  loadbalancer_id = "${ openstack_lb_loadbalancer_v2.lb_1.id }"
  region = "${var.region}"
}
#POOL 2
resource "openstack_lb_pool_v2" "pool_2" {
  name = "pool02"
  protocol    = "TCP"
  lb_method   = "ROUND_ROBIN"
  listener_id = "${ openstack_lb_listener_v2.listener_2.id }"
  # loadbalancer_id = "${ openstack_lb_loadbalancer_v2.lb_1.id }"
  region = "${var.region}"

  persistence {
    type        = "SOURCE_IP"
    # cookie_name = "null" #Another pull request test in progress (https://github.com/terraform-providers/terraform-provider-openstack/pull/58)
  }
}
```
Neutron shows error like this below and `terraform` kept creating pools instead of saying anything about this mismatch error.
```
2017/07/25 22:51:30 [DEBUG] plugin: terraform-provider-openstack:   "NeutronError": {
2017/07/25 22:51:30 [DEBUG] plugin: terraform-provider-openstack:     "detail": "",
2017/07/25 22:51:30 [DEBUG] plugin: terraform-provider-openstack:     "message": "Listener protocol HTTP and pool protocol TCP are not compatible.",
2017/07/25 22:51:30 [DEBUG] plugin: terraform-provider-openstack:     "type": "ListenerPoolProtocolMismatch"
2017/07/25 22:51:30 [DEBUG] plugin: terraform-provider-openstack:   }
2017/07/25 22:51:30 [DEBUG] plugin: terraform-provider-openstack: }
```

**Comment by jtopjian** [comment]
@sirajudheenam Thank you for reporting this!

**Comment by sirajudheenam** [comment]
I have tried after matching the `listener protocol type` to match `pool protocol type`. It still goes through the loop and created around 30 pools in 5 minutes, when load balancer provisioning status is `PENDING_UPDATE`. This happens particularly when we have more than one listeners and pools in the config and neutron takes a while to change the load balancer provisioning status to `ONLINE`.

eg.,

```hcl
#LISTENER 1
resource "openstack_lb_listener_v2" "listener_1" {
  name = "listener01"
  description = "This is the listener description"
  protocol        = "HTTP"
  protocol_port   = 8080
  loadbalancer_id = "${ openstack_lb_loadbalancer_v2.lb_1.id }"
  region = "${var.region}"
}

#POOL 1
resource "openstack_lb_pool_v2" "pool_1" {
  name = "pool01"
  protocol    = "HTTP"
  lb_method   = "ROUND_ROBIN"
  listener_id = "${ openstack_lb_listener_v2.listener_1.id }"
  region = "${var.region}"
  
  persistence {
    type        = "APP_COOKIE"
    cookie_name = "Cookie_test"
  }
}

#LISTENER 2
resource "openstack_lb_listener_v2" "listener_2" {
  name = "listener02"
  description = "This is the listener 2 description"
  protocol        = "TCP"
  protocol_port   = 8081
  loadbalancer_id = "${ openstack_lb_loadbalancer_v2.lb_1.id }"
  region = "${var.region}"
}

#POOL 2
resource "openstack_lb_pool_v2" "pool_2" {
  name = "pool02"
  protocol    = "TCP"
  lb_method   = "ROUND_ROBIN"
  listener_id = "${ openstack_lb_listener_v2.listener_2.id }"
  region = "${var.region}"

  persistence {
    type        = "SOURCE_IP"
  }
}
```

**Comment by jtopjian** [comment]
@sirajudheenam Thanks again. I think I'm starting to see what the problem might be. I could probably narrow it down a lot easier if you could provide the full output of:

```shell
TF_LOG=DEBUG OS_DEBUG=1 terraform apply
```

Could you post it to a Github gist?

**Comment by sirajudheenam** [comment]
Yes, Here it is.
https://gist.github.com/sirajudheenam/b78b6fa8ce1a626dd873d21bd7225733

**Comment by jtopjian** [comment]
@sirajudheenam Thanks. It looks like the gist has been cut off, though. It ends at the point where the pools are being created. Everything up to that point looks good, though. Is it possible to get the full log?

**Comment by sirajudheenam** [comment]
I have only removed the environment specific endpoint URLs in the gist and terminated `tf apply` command by Ctrl+C to avoid further creating Load Balancers in the environment.
It would be nice if `terraform` could wait until the loadbalancer's provisioning status is `ACTIVE` and State is `ONLINE` before it tries again than trying every 10 seconds.

**Comment by jtopjian** [comment]
@sirajudheenam Technically that is what it's doing. There should be no harm in sending the Create request when it is expecting a legitimate 409 reply. This is the same as saying "can't create, back off". If there were two separate calls: one to check for the state and one to create, that leaves room for a timing issue where the loadbalancer could to back into a PENDING state. Unless there is a bug, in which I would need to see the rest of the debug log to verify.

If your log is complete, then I think there is a miscommunication about the issue. What resource is being created repeatedly? Load balancers, listeners, or pools?

**Comment by sirajudheenam** [comment]
@jtopjian Thanks. The log is complete. It keeps creating pools.

**Comment by jtopjian** [comment]
@sirajudheenam My sincere apology - github was truncating the file. I never saw the message right in front of my face that said 

> This file has been truncated,

I'll take a look at this in its entirety and see if I can find anything.

**Comment by jtopjian** [comment]
@sirajudheenam I've just reviewed the rest of the logs and from what I can tell, API interactivity is correct. I have a feeling that pools are being created regardless of the 409 response being sent back. From reading the log, it looks like you're using an F5 load balancer as well as a vmware-based OpenStack environment. Unfortunately I have no experience with either of those. It's possible that they operate differently than a "vanilla" OpenStack environment or other load balancer drivers.

Can you try two things:

1. Set parallelism to 1:

```shell
$ terraform apply -parallelism=1
```

2. Create a dependency chain of all resources so that pool 2 depends on listener 2 which depends on pool 1 which depends on listener 1 which depends on the load balancer.

Both of this should effectively do the same thing: create the load balancer resources one at a time. This will remove any concurrency issues that might be causing the problem.

**Comment by jtopjian** [comment]
@sirajudheenam A patch was recently merged which might resolve this issue for you. If you are able to compile from source, can you test it out?

**Comment by fatmcgav** [comment]
@sirajudheenam We've just released v0.2.0 of this provider, which should resolve this issue. 

Feel free to test and raise a new issue if you encounter any problems. 
Closing this one in the meantime :)
